platform/Callback.h

```diff
@@ -35,209 +42,441 @@ namespace mbed {
+        // in preference to the universal constructor form.
+        copy(other);
     }
-
-    /** Create a Callback with a member function
-     *  @param obj      Pointer to object to invoke member function on
-     *  @param method   Member function to attach
-     */
-    template<typename T, typename U>
-    Callback(volatile U *obj, R(T::*method)(ArgTs...) volatile)
+#else // MBED_CONF_PLATFORM_CALLBACK_NONTRIVIAL
+    Callback(const Callback &other) = default;
+
+#ifndef __CC_ARM
+    Callback(Callback &&other) = default;
+#else
+    // ARMC5 doesn't support = default for move
+    Callback(Callback &&other) : CallbackBase(std::move(other))
     {
-        generate(method_context<volatile T, R(T::*)(ArgTs...) volatile>(obj, method));
     }
+#endif // CC_ARM
+#endif // MBED_CONF_PLATFORM_CALLBACK_NONTRIVIAL
 
     /** Create a Callback with a member function
      *  @param obj      Pointer to object to invoke member function on
      *  @param method   Member function to attach
      */
-    template<typename T, typename U>
-    Callback(const volatile U *obj, R(T::*method)(ArgTs...) const volatile)
+    template<typename Obj, typename Method, typename std::enable_if_t<mstd::is_invocable_r<R, Method, Obj, ArgTs...>::value, int> = 0>
+    Callback(Obj obj, Method method) : CallbackBase()
     {
-        generate(method_context<const volatile T, R(T::*)(ArgTs...) const volatile>(obj, method));
+        generate([obj, method](ArgTs... args) {
+            return detail::invoke_r<R>(method, obj, std::forward<ArgTs>(args)...);
+        });
     }
 
     /** Create a Callback with a static function and bound pointer
      *  @param func     Static function to attach
      *  @param arg      Pointer argument to function
      */
-    template<typename T, typename U>
-    Callback(R(*func)(T *, ArgTs...), U *arg)
+    template<typename Fn, typename BoundArg, typename std::enable_if_t<mstd::is_invocable_r<R, Fn, BoundArg, ArgTs...>::value, int> = 0>
+    Callback(Fn func, BoundArg arg) : CallbackBase()
     {
-        generate(function_context<R(*)(T *, ArgTs...), T>(func, arg));
+        generate([func, arg](ArgTs... args) {
+            return detail::invoke_r<R>(func, arg, std::forward<ArgTs>(args)...);
+        });
     }
 
-    /** Create a Callback with a static function and bound pointer
-     *  @param func     Static function to attach
-     *  @param arg      Pointer argument to function
+    // *INDENT-OFF*
+    /** Create a Callback with a function object
+     *  @param f Function object to attach
+     *  @note The function object is limited to a a few words of storage
      */
-    template<typename T, typename U>
-    Callback(R(*func)(const T *, ArgTs...), const U *arg)
+    template <typename F,
+            typename std::enable_if_t<
+                    !detail::can_null_check<F>::value &&
+                    mstd::is_invocable_r<R, F, ArgTs...>::value, int> = 0>
+    Callback(F f) : CallbackBase()
```

Comment:
Why not use a forward reference instead of a value ? of course, `std::forward` would have to be used in `generate`.

Reply:
Mainly because that's what it has done before, and it's what `std::function` does, and I don't want to deviate from that without good reason. That is specified as moving from the passed-by-value input.

I believe various C guidelines suggest that if the callee is always going to be retaining a copy, then passing by value is preferred.  The copy or move happens at the call-site.

Maybe we need to watch this 1-hour lecture:

https://stackoverflow.com/questions/17316386/pass-by-value-or-universal-reference
https://www.youtube.com/watch?v=PNRju6_yn3o&t=0s

(I haven't watched it yet...)

Or alternatively, just copy `std::function` and assume that it's okay. (Or at least "known").

Another article here which is relevant:

https://quuxplusone.github.io/blog/2019/03/27/design-space-for-std-function/

Reply:
Fair enough, following `std::function` interface is good practice in this case. 

In our case, the copy is _temporary_ as in the end it will be copied into the storage in `Callback`. I was thinking minimizing temporary copy would be a good thing. 

From my understanding, we end with the following operation (single layer of forwarding): 

|                             | lvalue argument | rvalue argument                                    |
|-----------------------------|-----------------|----------------------------------------------------|
| value parameter             | 1 copy + 1 move | 2 move (reduced to 1 move in case of move elision) |
| forward reference parameter | 1 copy          | 1 move                                             |
 
I'll have a look at the video when I got time. Josuttis is a good speaker.

Reply:
Table seems correct.

I don't think the standards folks are really thinking about the "move cost" for the actual `sizeof(F)`. I think they're thinking more of the "things hanging off F", like an allocated block - stuff like `string` or `vector`. So they want to minimise copies, but view any "move" as effectively free - so the two rows there are basically equivalent, getting it down to 1 copy.

I do think that in practice, we may get a code size benefit from not passing by value, particularly for GCC. We know that GCC is bad at passing things like `span` by value in A32/T32 mode, and I imagine that also extends to native pointers-to-member, which are 3 words big.

It's interesting that (for `std::function` and hence us too) `operator=` *does* take its parameter by forwarding reference. I can't immediately explain why that differs from the constructor.